webapp/channels/src/components/channel_bookmarks/channel_bookmarks_create_modal.tsx

```diff
@@ -0,0 +1,615 @@
+            return true;
+        }
+
+        if (emoji !== bookmark?.emoji) {
+            return true;
+        }
+
+        if (type === 'file') {
+            if (fileId && fileId !== bookmark?.file_id) {
+                return true;
+            }
+        }
+
+        if (type === 'link') {
+            return Boolean(link && link !== bookmark?.link_url);
+        }
+
+        return false;
+    })();
+    const isValid = (() => {
+        if (type === 'link') {
+            if (!link || linkError) {
+                return false;
+            }
+        }
+
+        if (type === 'file') {
+            if (!fileInfo || !displayNameValue || fileError) {
+                return false;
+            }
+        }
+
+        return true;
+    })();
+    const showControls = type === 'file' || (isValid || bookmark);
+
+    const cancel = () => {
+        if (type === 'file') {
+            uploadRequestRef.current?.abort();
+        }
+    };
+    const confirm = async () => {
+        setSaving(true);
+        if (type === 'link') {
+            const url = validHttpUrl(link);
+
+            if (!url) {
+                setSaveError(formatMessage(msg.linkInvalid));
+                return;
+            }
+
+            let validLink = url.toString();
+
+            if (validLink.endsWith('/')) {
+                validLink = validLink.slice(0, -1);
+            }
+
+            const {data: success} = await onConfirm({
+                image_url: icon,
+                link_url: validLink,
+                emoji,
+                display_name: displayNameValue,
+                type: 'link',
+            });
+
+            setSaving(false);
+
+            if (success) {
+                setSaveError('');
+                onHide();
+            } else {
+                setSaveError(formatMessage(msg.saveError));
+            }
+        } else if (fileInfo) {
+            const {data: success} = await onConfirm({
+                file_id: fileInfo.id,
+                display_name: displayNameValue,
+                type: 'file',
+                emoji,
+            });
+
+            if (success) {
+                setSaveError('');
+                onHide();
+            } else {
+                setSaveError(formatMessage(msg.saveError));
+            }
+        }
+    };
+
+    const confirmDisabled = saving || !isValid || !hasChanges;
+
+    return (
+        <GenericModal
+            enforceFocus={!showEmojiPicker}
+            keyboardEscape={false}
+            className='channel-bookmarks-create-modal'
+            modalHeaderText={formatMessage(bookmark ? msg.editHeading : msg.heading)}
+            confirmButtonText={formatMessage(bookmark ? msg.saveText : msg.addBookmarkText)}
+            handleCancel={(showControls && cancel) || undefined}
+            handleConfirm={(showControls && confirm) || undefined}
+            handleEnterKeyPress={(!confirmDisabled && confirm) || undefined}
+            onExited={handleOnExited}
```

Comment:
this functions get recreated every time that will cause `GenericModal` to re-render, specially when uploading a file as the progress updates.

Reply:
added useCallback